### PR TITLE
storage: Fix UAFs in PageStorage

### PR DESCRIPTION
**Pull request by breezewish**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tiflash/issues/7290

Problem Summary:

### What is changed and how it works?

`iter`, `iter->first`, `iter->second` will be all invalidated when iter is erased:

```c++
int main() {
    std::map<int, int> a;
    a.emplace(1, 2);
    auto it = a.find(1);
    a.erase(1);
    std::cout << it->second << std::endl;
    return 0;
}

==36182==ERROR: AddressSanitizer: heap-use-after-free on address 0x000105600cb0 at pc 0x0001025d5dac bp 0x00016d82b2b0 sp 0x00016d82b2a8
READ of size 4 at 0x000105600cb0 thread T0
    #0 0x1025d5da8 in main main.cpp:22
    #1 0x102971084 in start+0x200 (dyld:arm64+0x5084) (BuildId: fbb89662e6f23434b542f75185ac5e7432000000200000000100000000030c00)
    #2 0xb308fffffffffffc  (<unknown module>)

```

Thus we must only access iter inside a lock.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by breezewish** [comment]
Looks like we need to cherry pick these fixes for non-S3 versions as well, for example, when there is a GC, getByIDImpl may UAF. /cc @jayson-huang

**Review by lidezhu — APPROVED** [review]
LGTM

**Comment by JaySon-Huang** [comment]
> Looks like we need to cherry pick these fixes for non-S3 versions as well, for example, when there is a GC, getByIDImpl may UAF. /cc @JaySon-Huang

In theory, the page_id in MVCCMapType is only erased by the GC process. And the page_id accessed by getByIDImpl should be protected by the related `snapshot`, so the GC process will not invalidate the iterator.
But sure it is safer to get the k,v under lock instead of an iterator because we may break the assumption `the page_id in MVCCMapType is only erased by the GC process` one day.

**Review by JaySon-Huang — APPROVED** [review]
LGTM

File: dbms/src/Storages/Page/V3/PageDirectory.cpp [hunk 1]
```diff
@@ -1731,20 +1734,22 @@ PageDirectory<Trait>::getEntriesByBlobIds(const std::vector<BlobFileId> & blob_i
...
             if (iter == mvcc_table_directory.end())
                 return {blob_versioned_entries, total_page_size};
+            page_id = iter->first;
+            version_entries = iter->second;
         }
 
         while (true)
         {
             // `iter` is an iter that won't be invalid cause by `apply`/`gcApply`.
             // do scan on the version list without lock on `mvcc_table_directory`.
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/PageDirectory.cpp:1752]: remove these two lines

**Comment by breezewish** [comment]
> > Looks like we need to cherry pick these fixes for non-S3 versions as well, for example, when there is a GC, getByIDImpl may UAF. /cc @JaySon-Huang
> 
> In theory, the page_id in MVCCMapType is only erased by the GC process. And the page_id accessed by getByIDImpl should be protected by the related `snapshot`, so the GC process will not invalidate the iterator. But sure it is safer to get the k,v under lock instead of an iterator because we may break the assumption `the page_id in MVCCMapType is only erased by the GC process` one day.

Good to know. Let's only cherry back for S3.

**Comment by JaySon-Huang** [comment]
ref https://github.com/pingcap/tiflash/pull/7345 for backporting some fixes

**Pull request merged**